Pulp repo

sysadmind opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

pulp_repo
##### ANSIBLE VERSION

```
ansible 2.0.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY

This adds a module for working with pulp repos. http://www.pulpproject.org/. I added this under packaging/os because it seemed like the most relevant location for this module although it is not explicitly an os packaging module like others in this directory.

I created a class that lines up with a pulp server and added methods for getting and setting repo properties. I chose this design because it closely resembles the actual design of pulp and keeps the code logically separated into small methods. 

```
ok: [L-Pulp101.localdomain]

```

mscherer commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,278 @@
+            self.module.fail_json(msg="Error setting relative URL on repo.")
+
+    def set_repo_list(self):
+        cmd = self.pulp_command
```
Since that method is parsing output ( in parse_repo_list), you need to make sure that the command is run with C locales. See https://github.com/ansible/ansible-modules-core/pull/3257/files for a example.

resmo replied:
@mscherer wouldn't this be sufficient https://github.com/ansible/ansible-modules-extras/pull/1961/files#diff-6d7169a8a399575e4dfbcc7c8872ca83R232 ?

mscherer commented:
Seems check mode return a different thing that non check mode, I think it would be cleaner to get the same dict returned.

gregdek commented:
Thanks @sysadmind for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,280 @@
+    description:
+      - name of the repo to add or remove (correlates to pulp's repo-id)
+    required: true
+    default: null
```
`default: null` should be omitted if `required: true`

resmo commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,280 @@
+    description:
+      - Repo plugin type to use (i.e. rpm, docker)
+    required: true
+    default: null
```
`default: null` should be omitted if `required: true`

resmo commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,280 @@
+    choices: [ "present", "absent" ]
+  repo_type:
+    description:
+      - Repo plugin type to use (i.e. rpm, docker)
```
Not fully aware of pulp, is it possible to give the user a choice list for all possible repo types?

sysadmind replied:
Unfortunately the repo types are plugins to pulp, so they can be added or removed at any time and users can add custom types. Specifying choices would mean that they can't use this module on their own custom additions.

jtyr replied:
Add dot at the end of the sentence.

resmo commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,280 @@
+
+EXAMPLES = '''
+# Create a new repo with name 'my_repo'
+- pulp_repo: name=my_repo state=present
```
you defined `repo_type` to be required but in your examples it is missing.

resmo commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,280 @@
+'''
+
+RETURN = '''
+repo: Name of the repo that the action was performed on.
```
please have a look in other modules how to specify returns. e.g. https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/cloudstack/cs_account.py

gregdek commented:
Thanks @sysadmind for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+'''
+
+RETURN = '''
+---
```
This line should not be here.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+short_description: Add or remove Pulp repos from a remote host.
+description:
+   - Add or remove Pulp repos from a remote host.
+version_added: "2.1"
```
Should be changed to 2.2.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+options:
+  name:
+    description:
+      - name of the repo to add or remove (correlates to pulp's repo-id)
```
Change `name` to `Name` and `(correlates to pulp's repo-id)` to `(corresponds to pulp-id in Pulp)`. Add dot at the end of the sentence.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+    required: true
+  login_user:
+    description:
+      - User to connect to the pulp server as.
```
Change `pulp` to `Pulp`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+    default: null
+  login_password:
+    description:
+      - Password to connect to the pulp server with.
```
Change `pulp` to `Pulp`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+    default: null
+  relative_url:
+    description:
+      - Relative url for the local repository.
```
Change `url` to `URL`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+
+class pulp_server(object):
+    """
+    Class to interact with a pulp server
```
Change `pulp` to `Pulp`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+
+        else:
+            # Check to make sure all the settings are correct
+            if feed_url is not None:
```
Move the `check_mode` conditions to upper level:

```
if module.check_mode:
    changed = True
else:
    if feed_url is not None:
        ...
    if relative_url is not None:
        ...
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+                if not server.compare_repo_feed_url(repo, feed_url):
+                    if module.check_mode:
+                        module.exit_json(changed=True, repo=repo)
+                    server.set_repo_feed_url(repo, feed_url)
```
Change it to:

```
if not server.compare_repo_feed_url(repo, feed_url):
    if not module.check_mode:
        server.set_repo_feed_url(repo, feed_url)
    changed = True
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+            if module.check_mode:
+                module.exit_json(changed=True, repo=repo)
+            if server.create_repo(repo, feed_url):
+                changed = True
```
Change 259-262 to:

```
if not module.check_mode:
    server.create_repo(repo, feed_url)
changed = True
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+        if module.check_mode:
+            module.exit_json(changed=True, repo=repo)
+        if server.delete_repo(repo):
+            changed = True
```
Change 252-255 to:

```
if not module.check_mode:
    server.delete_repo(repo)
changed = True
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+                    if module.check_mode:
+                        module.exit_json(changed=True, repo=repo)
+                    server.set_repo_feed_url(repo, feed_url)
+                    changed = True
```
Change 268-271 to:

```
if not server.compare_repo_feed_url(repo, feed_url):
    if not module.check_mode:
        server.set_repo_feed_url(repo, feed_url)
    changed = True
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+                    if module.check_mode:
+                        module.exit_json(changed=True, repo=repo)
+                    server.set_repo_relative_url(repo, relative_url)
+                    changed = True
```
Change 275-278 to:

```
if not module.check_mode:
    server.set_repo_relative_url(repo, relative_url)
changed = True
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+            login_user=dict(default=None),
+            login_password=dict(default=None, no_log=True),
+            feed_url=dict(default=None),
+            relative_url=dict(default=None)
```
If `default` is `None`, you don't have to specify it. Example:

```
relative_url=dict()
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+
+EXAMPLES = '''
+# Create a new repo with name 'my_repo'
+- pulp_repo: name=my_repo repo_type=rpm state=present
```
Use pure YAML formatting for better legibility:

```
# Create a new repo with name 'my_repo'
- pulp_repo:
    name: my_repo
    repo_type: rpm state=present
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+- pulp_repo: name=my_repo repo_type=rpm state=present
+
+# Create a repo with a feed and a relative URL
+- pulp_repo: name=my_centos_updates repo_type=rpm feed_url='http://mirror.centos.org/centos/6/updates/x86_64/' relative_url='centos/6/updates' state=present
```
Use pure YAML formatting for better legibility:

```
- pulp_repo:
    name: my_centos_updates
    repo_type: rpm
    feed_url: http://mirror.centos.org/centos/6/updates/x86_64/
    relative_url: centos/6/updates
    state: present
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,285 @@
+- pulp_repo: name=my_centos_updates repo_type=rpm feed_url='http://mirror.centos.org/centos/6/updates/x86_64/' relative_url='centos/6/updates' state=present
+
+# Remove a repo from the pulp server
+- pulp_repo: name=my_old_repo repo_type=rpm state=absent
```
Use pure YAML formatting for better legibility:

```
- pulp_repo:
    name: my_old_repo
    repo_type: rpm
    state: absent
```

jtyr commented:
I would rather see an implementation which would call the [Pulp REST API](http://pulp.readthedocs.io/en/latest/dev-guide/integration/rest-api/) instead of the pulp-admin command. It makes the error handling and output parsing much simpler. Use the `fetch_url()` from the `ansible.module_utils.urls` for that. Also squash all the Git commits into one and do forced push to update your repo.

gregdek commented:
@sysadmind A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

sysadmind commented:
Thanks for the reminder. I will update with the necessary changes shortly.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+    name: my_centos_updates
+    repo_type: rpm
+    feed_url: 'http://mirror.centos.org/centos/6/updates/x86_64/'
+    relative_url: 'centos/6/updates'
```
You don't have to quote the two lines above.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+'''
+
+EXAMPLES = '''
+# Create a new repo with name 'my_repo'
```
What about to use this comment as a name of the task (and the same for the other examples)?

```
- name: Create a new repo with name 'my_repo'
  pulp_repo:
    name: my_repo
    repo_type: rpm
    state: present
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+DOCUMENTATION = '''
+---
+module: pulp_repo
+author: "Joe Adams <adams10301@gmail.com> github.com/sysadmind"
```
Modules normally have only the name of the author and his/her GitHub name here.

```
author: "Joe Adams (@sysadmind)"
```

Ansible developers usually use the GitHub name in comments to notify you about any updates of your module.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+        argument_spec=dict(
+            name=dict(required=True, aliases=['repo']),
+            state=dict(default="present", choices=["absent", "present"]),
+            repo_type=dict(required=True),
```
What about to make the `rpm` type as default?

```
repo_type=dict(default="rpm"),
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+    required: true
+  state:
+    description:
+      - The repo state
```
Missing dot at the end.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+    choices: [ "present", "absent" ]
+  repo_type:
+    description:
+      - Repo plugin type to use (i.e. rpm, docker).
```
Mark `rpm` and `docker` as parameters with the `C()` decoration.

```
- Repo plugin type to use (i.e. C(rpm), C(docker)).
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+            state=dict(default="present", choices=["absent", "present"]),
+            repo_type=dict(required=True),
+            login_user=dict(),
+            login_password=dict(no_log=True),
```
What about to set the user and password to "admin"  by default?

```
login_user=dict(default="admin"),
login_password=dict(default="admin", no_log=True),
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True, aliases=['repo']),
+            state=dict(default="present", choices=["absent", "present"]),
```
What about to add support for repo synchronization (`state: sync`) and publishing (`state: publish`)?

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+    if state == 'present':
+        if not repo_exists:
+            if not module.check_mode:
+                server.create_repo(repo, feed_url)
```
Here is missing parameter `relative_url`:

```
server.create_repo(repo, feed_url, relative_url)
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,291 @@
+            login_user=dict(),
+            login_password=dict(no_log=True),
+            feed_url=dict(),
+            relative_url=dict()
```
Here are missing additional parameters like `proxy_host` and `serve_http` and any other parameter provided by the `pulp-admin rpm repo create --help` command.

sysadmind commented:
I will spend some time updating the module to use the REST API and implement the pulp repo command more fully. I didn't want to add all the features in initially because I thought that might make this PR really big and this is my first module. I was hoping to improve and add features as time went on but that may not be the way of Ansible modules. I'll make some more changes and add some of the functionality like syncing and update the PR

gregdek commented:
@sysadmind Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

sysadmind commented:
ready_for_review
@jtyr Thanks for your feedback. I have gone back through and reworked the module to use the REST API for Pulp. I think that it's a good change, although it means that the module needs to handle a lot more logic. In making the changes, it did make it easy to add more parameters for importers and distributors. I'm hoping that this module is really close now.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,408 @@
+    required: true
+  state:
+    description:
+      - The repo state. A state of C(sync) will queue a sync of the repo. This is asynchronous but not delayed like a scheduled sync.
```
Please could you break the lines to be under 80 characters? The same applies to any line below.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,408 @@
+    )
+
+    # Use C locale since we are parsing the output when we call run_command()
+    module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C')
```
This can be now removed if you are using purely the REST API.

jtyr commented:
Great progress, @sysadmind. I have commented on couple of things and will comment on more when I have more time. In the meantime, please have a look at my [pulp_repos](https://github.com/jtyr/ansible-pulp_repos) role. I would like this module be able to do the same thing. I'm still missing the possibility to configure the distributors (e.g. `proxy_host`, `proxy_port`, `ssl_ca_cert`, `ssl_client_key`, `ssl_client_cert`, ...). But that should be an easy change for you now.

gregdek commented:
Thanks @sysadmind for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

sysadmind commented:
@jtyr I think that I have all of your requests integrated now. If you have any more, please let me know. Also I would appreciate if you could test the new `proxy_host`, `proxy_port`, `importer_ssl_ca_cert`, `importer_ssl_client_key`, and `importer_ssl_client_cert` parameters as I don't use them in my configuration so I don't know if they work as expected. I do see them show up properly in the importer config though so I feel pretty confident.

jtyr commented:
There is one last thing I'm missing in the module. When a repo is created via `pulp-admin` it automatically uses two distributors (`yum_distributor` and `expot_distributor`). I can see that you are trying to do the same. I found the `export_distributor` useless as it only creates ISO image with all the packages from the `yum_distributor` which only takes more space on the disk. Therefore I would like to see an option which would allow to specify which distributors should be used during the repo creation.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.urls import *
```
These imports should be on the top of the file together with the other imports.

sysadmind replied:
This is actually the standard in ansible modules according to their docs: 
http://docs.ansible.com/ansible/developing_modules.html#getting-your-module-into-ansible

jtyr replied:
It has changed in the recent version and now all imports can be on the top of the file.

sysadmind replied:
Do you have a link? I'm having trouble finding anything with my google-fu

jtyr replied:
I have discussed this recently here: https://github.com/ansible/ansible-modules-extras/pull/2369#issuecomment-223579676

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+            return True
+
+    def compare_repo_distributor_config(self, repo_id, **kwargs):
+        repo_config = self.get_repo_config_by_id(repo_id)
```
I would always make space between blocks of logic like between a variable definition and a `if`/`for` block:

```
repo_config = self.get_repo_config_by_id(repo_id)

for distributor in repo_config["distributors"]:
  ...
```

It improves readability. There is several places below where the same applies.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+                        return False
+        return True
+
+    def create_repo(self, repo_id, relative_url, feed_url=None, serve_http=False, serve_https=True, proxy_host=None, proxy_port=None):
```
Please wrap this line a bit. The best would be to keep it under 80 chars per line. Optionally, you can check your code with `pep8` command to see what else should be changed.

sysadmind replied:
I usually lint as I go in my editor with pep8/flake8. I wrapped this definition because it was starting to get pretty long, but I left many > 80 characters as is recommended for readability in the docs: http://docs.ansible.com/ansible/developing_modules.html#module-checklist

jtyr replied:
If you format it well, the readability improves by wrapping it at 80 chars. Too long lines are bad.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+        return True
+
+    def create_repo(self, repo_id, relative_url, feed_url=None, serve_http=False, serve_https=True, proxy_host=None, proxy_port=None):
+        url = self.host + '/pulp/api/v2/repositories/'
```
Instead of joining strings with `+`, I would use string formater:

```
url = '%s/pulp/api/v2/repositories/' % self.host
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+            data["notes"] = {
+                "_repo-type": "rpm-repo"
+            }
+        # self.module.fail_json(msg=data)
```
I guess this can be removed.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+            return True
+
+    def delete_repo(self, repo_id):
+        url = self.host + '/pulp/api/v2/repositories/%s/' % repo_id
```
Please  use only string formater:

```
url =  '%s/pulp/api/v2/repositories/%s/' % (self.host, repo_id)
```

The same repeats below.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+                    )
+        else:
+            data = dict()
+            data["id"] = publish_distributor
```
I have noticed inconsistency in quoting across your code. There is a convention that dictionary keys and identifiers should be single-quoted (e.g. `data['id'] = publish_distributor`, `method='POST'`, `if self.repo_type == 'rpm':`) and text messages should be double quoted (e.g. `msg="Failed to publish the repo the repo."`). It would be nice to unify this across your code.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,509 @@
+        if not repo_exists:
+            if not module.check_mode:
+                server.create_repo(
+                    repo,
```
I would name all parameters here:

```
if not module.check_mode:
  server.create_repo(
    repo_id=repo,
    relative_url=relative_url,
    feed_url=feed_url,
    serve_http=serve_http,
    serve_https=serve_https,
    proxy_host=proxy_host,
    proxy_port=proxy_port)

changed = True
```

Closing bracket on the same line as the last parameter. New line between the if block and the return command.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.urls import *
```
Please move the import to the top of the file. Here are more details: https://github.com/ansible/ansible-modules-extras/pull/2369#issuecomment-223579676

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+
+
+
+
```
Please keep only 2 blank lines here.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+            yum_distributor['distributor_config']['relative_url'] = relative_url
+            data['distributors'].append(yum_distributor)
+
+            export_distributor = dict()
```
Please could you make the list of distributors optional? See https://github.com/ansible/ansible-modules-extras/pull/1961#issuecomment-228165780.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+author: "Joe Adams (@sysadmind)"
+short_description: Add or remove Pulp repos from a remote host.
+description:
+   - Add or remove Pulp repos from a remote host.
```
Please remove one space at the beginning of this line to have the number of spaces dividable by 2.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+    required: false
+  force_basic_auth:
+    description:
+      - httplib2, the library used by the uri module only sends authentication
```
`uri` is name of module, so it should be `M(uri)`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,556 @@
+    default: null
+  importer_ssl_client_key:
+    description:
+      - Private key to the certificate specified in importer_ssl_client_cert,
```
`importer_ssl_client_cert` should be `I(importer_ssl_client_cert)`.

jtyr replied:
Actually I'm not sure here if it should be `I()` or `C()`. According to the [documentation](http://docs.ansible.com/ansible/developing_modules.html#module-checklist):

> U() for urls, C() for files and options, I() for params, M() for modules

But I'm not sure what's _option_ and what's _parameter_. In the [`yum_repository`](https://github.com/ansible/ansible-modules-extras/blob/devel/packaging/os/yum_repository.py) module, I'm using `I()` for `name`, `state`, ... and `C()` for the values.

I would say that _parameter_ is the name of something and _option_ is the list of values which I can select from. But English is not my mother tongue so I can be wrong.

sysadmind commented:
I added an option `add_export_distributor` to disable that only for RPM repos. I'm not sure how to realistically allow adding/disabling distributors across the board because other repo types can have other distributors. If you think there's a better way to do it, please let me know.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,562 @@
+    required: false
+    default: 'yes'
+    choices: ['yes', 'no']
+'''
```
Before the end of the `DOCUMENTATION` block, there should be `notes` describing limitations of this module. I would just mention that the module is currently capable to create YUM repos only:

```
notes:
  - This module can currently create only YUM repositories. Contributions to support other repo types are welcome.
```

sysadmind commented:
I updated the notes. I changed the wording to show the limitation is in the importers and distributors since the API doesn't seem to impose types until those are configured.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+  description: Name of the repo that the action was performed on.
+  returned: success
+  type: string
+  sample: repo_name
```
I would change this to `my_repo` to have it in line with the example above.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+  url_password:
+    description:
+        - The password for use in HTTP basic authentication to the pulp API.
+          If the C(url_username) parameter is not specified, the C(url_password)
```
Please use `I()` for parameters.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+        used on personally controlled sites using self-signed certificates.
+    required: false
+    default: 'yes'
+    choices: ['yes', 'no']
```
Please add spaces around the parentheses and use double quotes in order to have it the same across the all DOCUMENTATION block:

```
choices: [ "yes", "no" ]
```

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+    description:
+      - Name of the repo to add or remove. This correlates to repo-id in Pulp.
+    required: true
+  state:
```
Please could you order all the options alphabetically for better readability?

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+    argument_spec = url_argument_spec()
+    argument_spec.update(
+        name=dict(required=True, aliases=['repo']),
+        state=dict(default="present", choices=["absent", "present", "sync", "publish"]),
```
Please change the double quotes to single quotes to keep the same style.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+
+        else:
+            # Check to make sure all the settings are correct
+            # The importer config gets overwritten on set and not updated, so we set the whole config at the same time.
```
Please could you wrap this like? It would be really great if you could keep all lines in the document under 80 characters  to be fully PEP8 compliant.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,568 @@
+                changed = True
+
+    module.exit_json(changed=changed, repo=repo)
+
```
I would add one more space here ;o)

sysadmind commented:
I made some more changes based on your recommendations.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,595 @@
+def main():
+    argument_spec = url_argument_spec()
+    argument_spec.update(
+        name=dict(required=True, aliases=['repo']),
```
Please order also this list alphabetically as you did with the documentation.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,595 @@
+        argument_spec=argument_spec,
+        supports_check_mode=True)
+
+    repo = module.params['name']
```
This could be ordered alphabetically as well.

jtyr commented:
I have added two more comments. Some lines are still not under 80 chars (please run `pep8 pulp_repo.py`) but otherwise I'm pretty happy how the module looks. Now I just need to find time to test it ;o)

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,595 @@
+        C(publish) will use the repository's distributor to publish the content.
+    required: false
+    default: present
+    choices: [ "present", "absent", "sync", "publish" ]
```
You don't have to quote the values here.

```
choices: [ present, absent, sync, publish ]
```

sysadmind commented:
I reordered the arguments and the variables after the arguments. I did change one more line in terms of line length, but the remaining lines are all just over 80 characters and I believe that changing them will just make the code harder to read so I'm going to leave them like they are. The contributing docs explicitly say that it's not a hard limit and that line length limits can actually hurt readability. 

As far as quoting values in the documentation, I'm going to leave the quotes. Looking through other modules in the repo, they are fairly consistently quoted. I want to keep the styling similar to the rest of the repo for easier readability and consistency.

jtyr commented:
> the remaining lines are all just over 80 characters and I believe that changing them will just make the code harder to read

I normally aim for PEP8 compliant code but your reason is fair enough.

Regarding the quoting - the `DOCUMENTATION` is suppose to be YAML and in YAML it's not necessary to quote all values. Most people don't know that and quote all values like in JSON. I would recommend to remove quotes everywhere where it's nor necessary. Keep quotes around `yes`/`true`, `no`/`false` and `null` as they would be turned to `True`, `False` and `None` once read by Python. And of course keep it around values beginning at a special character (e.g. `-`, `[`, `{`, `!`, ...). You can find more details [here](https://en.wikipedia.org/wiki/YAML#Syntax). I think it's good to know this not only for the module development but also for Ansible as it's using YAML everywhere. Good usage of YAML can improve readability of the Ansible role/playbook significantly.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+                distributor_url,
+                data=json.dumps(data),
+                method='PUT'
+            )
```
I would consider to move the ending parenthesis to the last like of code to keep it the same like in the case of `self.module.fail_json(...)` or `fetch_url(...)`.

```
response, info = fetch_url(
  self.module,
  distributor_url,
  data=json.dumps(data),
  method='PUT')
```

There is several other occurrences elsewhere in the code.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+                url,
+                data=json.dumps(data),
+                method='POST')
+            if info['status'] != 202:
```
Please add new line above.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+                status_code=info['status'],
+                response=info['msg'],
+                url=url)
+
```
Please remove this empty line.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+    def check_repo_exists(self, repo_id):
+        try:
+            repo_config = self.get_repo_config_by_id(repo_id)
+
```
Please remove this empty line and also three lines lower above the `else:`.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+    if state == 'absent' and repo_exists:
+        if not module.check_mode:
+            server.delete_repo(repo)
+        changed = True
```
Please add new line above.

jtyr commented on packaging/os/pulp_repo.py:
```diff
@@ -0,0 +1,597 @@
+                    add_export_distributor=add_export_distributor)
+
+            changed = True
+
```
Please remove this empty line.

sysadmind commented:
I take your point about that YAML. I fixed it and updated the spacing in the code per your recommendations.

sysadmind commented:
I pushed an update that cleaned up some quoting and added a wait_for_completion option to wait for async tasks to complete before returning.

sysadmind commented:
Not sure why the build is failing. Looks like it might be an upstream issue.

jtyr commented:
You probably need to rebase your tree. Something like this:

```
git checkout devel
git pull
git checkout pulp_repo
git rebase devel
git push -f
```

sysadmind commented:
I didn't think of updating my source branch. Good call. Thanks @jtyr

gregdek commented:
The pending action label clearly does not apply here. 

@jtyr you've commented a ton on this one, so if you can give a +1 at this point, we can fast-track it into the new modules meeting for inclusion.

jtyr commented:
@gregdek I just tried to use this module and unfortunately the examples did not work for me. I did not dig into it too much but first thing I have noticed is that a lot of parameters which are defined in the documentation are missing in the `argument_spec.update()` in the `main()`. For example `url_username` and `url_password` params are missing which are required for logging in to the Pulp server. If @sysadmind will fix it, I can do run tests.

@gregdek I would appreciate if you could fast-track the `jenkins_plugin` module (https://github.com/ansible/ansible-modules-extras/pull/1730) into the new modules meeting for inclusion (it got one +1).

sysadmind commented:
@jtyr: Those arguments come from `argument_spec = url_argument_spec()` They are handled by Ansible apart from my module. Those arguments get passed through to the `fetch_url` calls automatically. 

If you were having trouble with the authentication, did you try the `force_basic_auth` parameter? I've noticed that my instance of Pulp requires that to be set so I'm not sure if that's more universally true or a limitation of my environment.

jtyr commented:
I did some more tests and have found out this:
- The Pulp API runs via HTTPS only so you need to change the default `pulp_host` to `https://127.0.0.1`.
- It doesn't make much sense to put `pulp_host` into the example as the code runs locally on the Pulp server anyway.

Otherwise I see your point about the `url_argument_spec()` now. I can imagine that people will need to set certain parameters for all their `pulp_repo` tasks. It would be handy to have the possibility to predefine such set of parameters as a variable and then use it within all tasks without the need to copy them all every time (if you need to add a param, you can add it on one place instead of on 100 places). I'm doing that in the [`yum_repository`](https://github.com/ansible/ansible-modules-extras/blob/devel/packaging/os/yum_repository.py#L689) and [`jenkins_plugin`](https://github.com/ansible/ansible-modules-extras/pull/1730/files#diff-e24f65156cc9b7c1dfe44242891e885aR738) by using the `params` option. The usage for Pulp could be as follows:

```
---

- name: Pulp server repos
  hosts: pulp01
  vars:
    pulp_params:
      validate_certs: no
      url_username: admin
      url_password: admin
  tasks:
    - name: Create local repo
      pulp_repo:
        name: repo1
        params: "{{ pulp_params }}"
    - name: Create repo with feed
      pulp_repo:
        name: repo2
        feed_url: http://some.server.com/path/to/the/repo
        params: "{{ pulp_params }}"
```

The last but not least, the module should be called `pulp_repository` to follow the convention set by other modules (e.g. `apt_repository`, `yum_repository`, `zypper_repository`).

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,644 @@
+    description:
+      - Name of the repo to add or remove. This correlates to repo-id in Pulp.
+    required: true
+  proxy_host:
```
What about the change this to `proxy_server`?

jtyr replied:
I realize you you follow the naming convention given by `pulp-admin` where is the `--proxy-host` option. Not sure what's better - `host` or `server`?

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,644 @@
+      - Distributor to use when state is C(publish). The default is to
+        publish all distributors.
+    required: false
+  pulp_host:
```
What about the change this to `pulp_server`?

jtyr replied:
I would change the postfix to whatever you decide to use with the case of `proxy_host`.

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,644 @@
+      - Whether or not to add the export distributor to new C(rpm) repositories.
+    required: false
+    default: false
+  feed_url:
```
I would rename this to `feed` to follow the convention given by `pulp-admin` with its `--feed`.

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,644 @@
+        state=dict(
+            default="present",
+            choices=['absent', 'present', 'sync', 'publish']),
+        wait_for_completion=dict(type="bool", default=False))
```
Please could you swap the `type` with the `default` parameters to have them in the same order as in the cases above?

gthieleb commented:
I have written my own python script to perform pulp repo creation, syncing and copying repos, which uses pulp-admin behind the scenes. But having an ansible module for pulp administration would be  :+1: .

Do you think there would be a place in the module to support copy tasks? What about creating the repo config files for the client side?

sysadmind commented:
@gthieleb I have a separate module that I have developed to copy content units between repositories. I think that it makes sense to keep units and repositories separate. It may also make sense to create a client side module for that task. Once the pulp_repo module is merged, I'll work on getting the other modules contributed upstream.

dvanallen commented:
I would love to use this module and it looks like you guys have put a lot of work into it - anything I can do to help get this merged?

jtyr commented:
@dvanallen You can test the module and comment with the magic word as explained [here](https://github.com/ansible/ansible-modules-extras/pull/1961#issuecomment-205521506).

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,643 @@
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.urls import *
```
It would be better to have explicit imports like this:

```
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url
from ansible.module_utils.urls import url_argument_spec
```

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,643 @@
+import json
+from time import sleep
+
+# import module snippets
```
This comment is not necessary as the bellow are not snippets (it used to be in <v2.x) but regular modules.

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,643 @@
+
+    def check_repo_exists(self, repo_id):
+        try:
+            repo_config = self.get_repo_config_by_id(repo_id)
```
The variable assignment is not necessary as you don't use the variable anywhere else. Just call the function itself:

```
try:
    self.get_repo_config_by_id(repo_id)
```

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,643 @@
+    def check_repo_exists(self, repo_id):
+        try:
+            repo_config = self.get_repo_config_by_id(repo_id)
+
```
There should be no space here.

jtyr commented on packaging/pulp_repo.py:
```diff
@@ -0,0 +1,643 @@
+
+        except IndexError:
+            return False
+
```
The same as above - no space here.

sysadmind commented:
Yes @dvanallen, testing would be a HUGE help as many different users have different use cases for pulp and it's hard to test for them all. Please report back with any issues or the magic words as @jtyr links to above to help get this merged.

Nosmoht commented:
shipit

If someone needs code to test it, here is how i did it on Vagrant. Comment the roles as you might not have them

```
---
- hosts: all
  become: true
  become_method: sudo
  vars:
    pulp_repos:
    - name: CentOS-7-x86_64
      feed: http://mirror.centos.org/centos/7/os/x86_64/
      relative_url: centos/7/base
    - name: CentOS-7-x86_64-updates
      feed: http://mirror.centos.org/centos/7/updates/x86_64/
      relative_url: centos/7/updates
    - name: WANdisco-CentOS-7-x86_64-git
      feed: http://opensource.wandisco.com/centos/7/git/x86_64/
      relative_url: wandisco/centos/7/git
    - name: ELRepo-7-x86_64-kernel
      feed: http://elrepo.org/linux/kernel/el7/x86_64/
      relative_url: elrepo/7/kernel
  roles:
  - role: epel
  - role: mongodb
  - role: qpid
  - role: apache
  - role: pulp
  post_tasks:
  - name: Ensure repo
    pulp_repo:
      name: '{{ item.name }}'
      repo_type: rpm
      feed: '{{ item.feed }}'
      relative_url: '{{ item.relative_url }}'
      url_username: admin
      url_password: admin
      validate_certs: false
      force_basic_auth: true
      state: '{{ item.state | default(omit) }}'
    with_items: '{{ pulp_repos }}'
    tags: pulp_repos
```

coreaut commented:
@sysadmind you could check if the **importer_ssl_ca_cert** parameter is multiline. If not treat it as a file path to the certs?

edit: or use file lookup. Not sure what is more the "ansible way"? (Module supports paths or require file lookups)

jtyr commented:
The `lookup` plugin is only for local files. You want to read content of remote file which ca be done only via extra task. I think that it would be nice to be able to define remote path in the `importer_ssl_*` vars. Just check if the content of the var is actually a path to a file and if yes then read the content.

coreaut commented:
@jtyr only **importer_ssl_ca_cert** needs to be a string. The other **importer_ssl_**\* vars already have to be file paths.
Thanks for the **lookup** hint :smile:

coreaut commented:
shipit

@sysadmind I fully switched to your module (replaced it with my own).
Everything worked and I encountered no bugs so far (I have 33 repos). 

The only nuisance is the extra Task required for **importer_ssl_ca_cert** :smile:.

sysadmind commented:
@Nebelwolf I'll try and get around to updating it to work with your ssl ca certs. It's hard for me to test because I don't use that setting. 

If I understand correctly, it can either be a file path or actual certificate content (multiline). Is that the case? If I can figure out those changes and get them pushed to a branch, would you be able to test? Or do you have a testing procedure (like with a pulp docker container) that I could use for testing this functionality so that I can confirm I fix it for you?

coreaut commented:
@sysadmind 
For example currently I have to do something like this:

```
  - shell: cat "{{pulp_redhat_feed_cacert}}"
    register: datacacert
    changed_when: false

  - name: create RedHat repos
    pulp_repo:
      url_username: admin
      url_password: "{{ pulp_admin_password }}"
      force_basic_auth: true
      name: "{{ item['name'] }}"
      feed: "{{ item['feed'] }}"
      pulp_host: {{ pulp_host_url }}
      relative_url: "{{ item['relative_url'] }}"
      importer_ssl_ca_cert: "{{ datacacert.stdout }}"
      importer_ssl_client_cert: "{{ pulp_redhat_feed_cert }}"
      importer_ssl_client_key: "{{ pulp_redhat_feed_key }}"
    with_items: "{{ pulp_repo_map.redhat }}"
    notify:
      - resync repos
```

You see the extra task for **importer_ssl_ca_cert**? I just meant it would be nice if I could just give it the file path as parameter. Now it needs the multiline var with the cert.

So you would need to read the cert contents from the file path in your module code and pass that to the api call.

Maybe you should accept only paths for **importer_ssl_ca_cert**? Would be more consistent with the other importer_ssl_\* params?

I can test it when you change something. Sorry theres no docker container for this. You need an valid RedHat subscription to test it.

sysadmind commented:
I think I have this working as you would expect @Nebelwolf and @jtyr. I was able to curl the api and see the certs/keys by both entering the content and specifying a path to a file. Can you please test to make sure it works on your systems?

coreaut commented:
@sysadmind sorry for the delay. Yes it works now as expected thanks!

sysadmind commented:
@Nebelwolf @Nosmoht @dvanallen @jtyr If there is no other feedback, I would appreciate your assistance in getting this merged. You can comment on this PR as stated in https://github.com/ansible/ansible-modules-extras/pull/1961#issuecomment-205521506 and that will mark this PR for merge by the maintainers.  I appreciate your help and feedback in this PR.

coreaut commented:
@sysadmind do I have to comment the "magic word" again?
 [my previous comment](https://github.com/ansible/ansible-modules-extras/pull/1961#issuecomment-258135973)

shipit :boat:

sysadmind commented:
Moving to core repo.
